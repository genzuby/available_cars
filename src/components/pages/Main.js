import React from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import useInitalLoading from '../../hooks/useInitalLoading';
import Filters from '../Content/Filters';
import AvailableCars from '../Content/AvailableCars';
import { mainStyle } from '../../style/componentStyle';

function Main() {
  useInitalLoading();

  return (
    <Container fluid style={mainStyle} data-testid="main">
      <Row>
        <Col style={{ maxWidth: '400px' }}>
          <Filters />
        </Col>
        <Col style={{ margin: '0 48px 0 0' }}>
          <AvailableCars />
        </Col>
      </Row>
    </Container>
  );
}

export default Main;
