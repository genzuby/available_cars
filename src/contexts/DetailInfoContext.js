import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getDetailInfo } from '../redux/actions';

export const DetailInfoContext = React.createContext();

export function DetailInfoProvider({ children }) {
  const [detailInfo, setDetailInfo] = useState({});
  const carDetailInfo = useSelector(state => state.carDetailInfo);
  const dispatch = useDispatch();

  useEffect(() => {
    setDetailInfo(carDetailInfo);
  }, [carDetailInfo]);

  const fetchDetailInfo = stockNumber => {
    dispatch(getDetailInfo(stockNumber));
  };

  return (
    <DetailInfoContext.Provider value={{ detailInfo, fetchDetailInfo }}>
      {children}
    </DetailInfoContext.Provider>
  );
}
