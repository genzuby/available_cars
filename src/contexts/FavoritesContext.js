import React from 'react';
import { useSelector } from 'react-redux';
import useLocalStorage from '../hooks/useLocalStorage';

export const FavoritesContext = React.createContext();

const LOCAL_STOAGE_KEY = 'AUTO1_MY_FAVORITE_CAR';

export function FavoritesProvider({ children }) {
  const [favorites, setFavorites] = useLocalStorage(LOCAL_STOAGE_KEY, []);
  const carDetailInfo = useSelector(state => state.carDetailInfo);

  const addFavorite = () => {
    setFavorites(prevFavorites => {
      let checkDuplication;
      if (prevFavorites !== 0) {
        checkDuplication = prevFavorites.filter(
          item => item.stockNumber === carDetailInfo.stockNumber
        );
      }

      return checkDuplication.length !== 0
        ? prevFavorites
        : [carDetailInfo, ...prevFavorites];
    });
  };

  const deleteFavorite = (stockNumber: string) => {
    setFavorites(prevFavorites => {
      const filteredFavorites = prevFavorites.filter(
        item => item.stockNumber !== +stockNumber
      );

      return filteredFavorites;
    });
  };

  const value = { favorites, addFavorite, deleteFavorite };

  return (
    <FavoritesContext.Provider value={value}>
      {children}
    </FavoritesContext.Provider>
  );
}
