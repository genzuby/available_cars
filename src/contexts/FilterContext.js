import React, { useState } from 'react';

export const FilterContext = React.createContext();

export function FilterProvider({ children }) {
  const [filter, setFilter] = useState({});

  return (
    <FilterContext.Provider value={{ filter, setFilter }}>
      {children}
    </FilterContext.Provider>
  );
}
