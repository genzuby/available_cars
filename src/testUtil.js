import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

export const providerValue = {
  color: 'red',
  fuelType: 'Diesel',
  manufacturerName: 'Dodge',
  mileage: { number: 138210, unit: 'km' },
  modelName: 'Caliber',
  pictureUrl: 'car.svg',
  stockNumber: 10907,
};

const initialState = {
  carDetailInfo: providerValue,
  manufacturersFilter: [{ name: 'Audi' }],
  colorFilter: ['red'],
  carList: { cars: [providerValue], totalPageCount: 10, totalCarsCount: 100 },
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

export const store = mockStore(initialState);

export const localStorageMock = {
  getItem: jest.fn().mockImplementation(key => localStorageItems[key]),
  setItem: jest.fn().mockImplementation((key, value) => {
    localStorageItems[key] = value;
  }),
  clear: jest.fn().mockImplementation(() => {
    localStorageItems = {};
  }),
  removeItem: jest.fn().mockImplementation(key => {
    localStorageItems[key] = undefined;
  }),
};

export let localStorageItems = {};
