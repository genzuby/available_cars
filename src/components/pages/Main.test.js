import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Main from './Main';
import { FavoritesProvider } from '../../contexts/FavoritesContext';
import { FilterProvider } from '../../contexts/FilterContext';
import { Provider } from 'react-redux';
import { store } from '../../testUtil';
import {
  getColorsList,
  getManufaturersList,
  getAvailableCarList,
} from '../../redux/actions';

beforeEach(async () => {
  await store.dispatch(getColorsList());
  await store.dispatch(getManufaturersList());
  await store.dispatch(getAvailableCarList({ color: 'red' }));

  render(
    <Provider store={store}>
      <FavoritesProvider value={{ favorites: {} }}>
        <FilterProvider value={{ color: 'red' }}>
          <Main />
        </FilterProvider>
      </FavoritesProvider>
    </Provider>,
    { wrapper: MemoryRouter }
  );
});

it('snapshot', () => {
  expect(screen.getByTestId('main')).toMatchSnapshot();
});

it('renders with context value and redux store', () => {
  expect(screen.getByTestId('main')).toBeInTheDocument();
});
