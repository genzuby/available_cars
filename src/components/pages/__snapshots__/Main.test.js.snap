// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="container-fluid"
  data-testid="main"
  style="margin: 80px 0px; padding: 24px 48px 0px 24px; display: grid; min-height: 76vh;"
>
  <div
    class="row"
  >
    <div
      class="col"
      style="max-width: 400px;"
    >
      <div
        class="input-group"
        data-testid="filters"
        style="border: 2px solid #ededed; padding: 24px; margin: 24px; display: flex;"
      >
        <div
          class="sc-pFZIQ duAHLa"
          data-testid="dropdown"
        >
          <p
            class="sc-jrAGrp eEGqha"
          >
            color
          </p>
          <div
            class="sc-iBPRYJ fNSIpf"
          >
            <button
              class="sc-gKsewC bdczbD"
              data-testid="dropdown-button"
            >
              All car colors
              <span
                class="--dropdown-arrow"
              />
            </button>
            <div
              class="list-group"
              data-testid="dropdown-list"
              style="width: 100%; position: absolute; z-index: 2; background-color: rgb(255, 255, 255); cursor: pointer; left: 0px; top: 48px; display: none;"
            >
              <button
                class="list-group-item list-group-item-action"
              >
                red
              </button>
            </div>
          </div>
        </div>
        <div
          class="sc-pFZIQ duAHLa"
          data-testid="dropdown"
        >
          <p
            class="sc-jrAGrp eEGqha"
          >
            manufacturers
          </p>
          <div
            class="sc-iBPRYJ fNSIpf"
          >
            <button
              class="sc-gKsewC bdczbD"
              data-testid="dropdown-button"
            >
              All manufacturers
              <span
                class="--dropdown-arrow"
              />
            </button>
            <div
              class="list-group"
              data-testid="dropdown-list"
              style="width: 100%; position: absolute; z-index: 2; background-color: rgb(255, 255, 255); cursor: pointer; left: 0px; top: 48px; display: none;"
            >
              <button
                class="list-group-item list-group-item-action"
              >
                Audi
              </button>
            </div>
          </div>
        </div>
        <div
          class="sc-fubCfw bCEQtI"
        >
          <button
            class="sc-gsTCUz dUSUzg"
            data-testid="filter-button"
          >
            Filter
          </button>
        </div>
      </div>
    </div>
    <div
      class="col"
      style="margin: 0px 48px 0px 0px;"
    >
      <div
        class="sc-hKgILt kUrLvW"
        data-testid="availableCarList"
      >
        <h3
          class="--list-title"
        >
          Available cars
        </h3>
        <p
          class="--list-desc"
        >
          Showing 10 of 
          100
           results
        </p>
        <div
          data-testid="carList"
        >
          <div
            class="card"
            data-testid="itemCard"
            style="margin-bottom: 8px; width: 100%; position: relative;"
          >
            <div
              class="card-body"
            >
              <div
                class="sc-eCssSg jQmxdX"
              >
                <img
                  alt="Dodge Caliber"
                  class="sc-jSgupP htlqyq"
                  src="car.svg"
                  width="120px"
                />
                <div>
                  <div
                    class="card-title h5"
                    style="font-size: 24px; font-weight: bold;"
                  >
                    Dodge Caliber
                  </div>
                  <p
                    class="card-text"
                  >
                    Stock #10907 - 138210 km - Diesel
                  </p>
                  <a
                    href="/10907"
                  >
                    <div
                      class="sc-dlfnbm jkoPfC"
                    >
                      View Details
                    </div>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-kEjbxe eUZXih"
          data-testid="pagination"
        >
          <button
            class="sc-crrsfI jomiL"
            disabled=""
          >
            First
          </button>
          <button
            class="sc-crrsfI jomiL"
            disabled=""
          >
            Previous
          </button>
          <p
            class="sc-iqHYGH coxDri"
            data-testid="pagination-view"
          >
            Page 
            1
             of 
            10
          </p>
          <button
            class="sc-crrsfI jomiL"
          >
            Next
          </button>
          <button
            class="sc-crrsfI jomiL"
          >
            Last
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
